--- a/folia-api/build.gradle.kts
+++ b/folia-api/build.gradle.kts
@@ -105,17 +_,21 @@
         java {
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../folia-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../folia-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
+            srcDir(file("../folia-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
+            srcDir(file("../folia-api/src/test/resources"))
         }
     }
 }
@@ -259,3 +_,12 @@
 tasks.check {
     dependsOn(scanJarForBadCalls)
 }
+
+// Kitin start - Hide unnecessary warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// Kitin end - Hide unnecessary warnings
+
+dependencies {
+    compileOnly("com.google.code.findbugs:jsr305:3.0.2")
+}
