From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Sat, 20 Dec 2025 20:22:57 +0800
Subject: [PATCH] FoliaFix_WaypointManager


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 455e4f95efc0c27e09d41d4a4deff5ab913433ee..755fe00ea86a57aa899c19316d3dd90900d8e973 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -61,6 +61,7 @@ import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.ServerScoreboard;
 import net.minecraft.server.players.SleepStatus;
+import me.sucixr.kitin.fixes.foliafix.KitinServerWaypointManager;
 import net.minecraft.server.waypoints.ServerWaypointManager;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundSource;
@@ -649,7 +650,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         this.sleepStatus = new SleepStatus();
         this.gameEventDispatcher = new GameEventDispatcher(this);
         this.randomSequences = Objects.requireNonNullElseGet(randomSequences, () -> this.getDataStorage().computeIfAbsent(RandomSequences.TYPE));
-        this.waypointManager = new ServerWaypointManager();
+        this.waypointManager = new KitinServerWaypointManager(this);
         this.environmentAttributes = EnvironmentAttributeSystem.builder().addDefaultLayers(this).build();
         //this.updateSkyBrightness(); // Folia - region threading - delay until first tick
         // Paper start - rewrite chunk system
