From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 10:58:09 +0800
Subject: [PATCH] Perf_Entity-push


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 13f23918b6a9db52742a01bf16c69389a1c16dc9..ec4d3f6cbf659090da515f6db89478e1aa1cd77b 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2334,6 +2334,14 @@ public abstract class Entity
         if (!this.isPassengerOfSameVehicle(entity)) {
             if (!entity.noPhysics && !this.noPhysics) {
                 if (this.level.paperConfig().collisions.onlyPlayersCollide && !(entity instanceof ServerPlayer || this instanceof ServerPlayer)) return; // Paper - Collision option for requiring a player participant
+                //Kitin start
+                if (me.sucixr.kitin.config.KitinConfig.useSimplerEntityPush) {
+                    if (me.sucixr.kitin.performance.physics.SocialBubbleEngine.getInstance().shouldSkipHardCollision(this,entity)) {
+                        return;
+                    }
+                    return;
+                }
+                //Kitin end
                 double xa = entity.getX() - this.getX();
                 double za = entity.getZ() - this.getZ();
                 double dd = Mth.absMax(xa, za);
