From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Thu, 25 Dec 2025 13:23:09 +0800
Subject: [PATCH] OPT_ChunkLazyLoading


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index d94c0d15f62b67378669d4ce7252d99b8f743fa5..78532e1e4d3eba6c58af93b8ae9cb5c74d1adb56 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -908,6 +908,11 @@ public final class RegionizedPlayerChunkLoader {
         }
 
         void update() {
+            // Kitin start
+            if (me.sucixr.kitin.scheduler.controller.ChunkLoadController.shouldSkipUpdate(this.player)) {
+                return;
+            }
+            // Kitin end
             TickThread.ensureTickThread(this.player, "Cannot update player asynchronously");
             if (this.removed) {
                 throw new IllegalStateException("Updating removed player chunk loader");
