From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Thu, 25 Dec 2025 13:23:09 +0800
Subject: [PATCH] OPT_ChunkLazyLoading


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index d94c0d15f62b67378669d4ce7252d99b8f743fa5..ca023f9afeb4b8697664e2f5460d402c324084f8 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -908,6 +908,12 @@ public final class RegionizedPlayerChunkLoader {
         }
 
         void update() {
+            // Kitin start - lazy chunk update gate (with init guard)
+            if (this.lastSendDistance != Integer.MIN_VALUE // 或者 lastChunkX != MIN_VALUE 也行
+                    && me.sucixr.kitin.network.misc.ChunkLoadController.shouldSkipUpdate(this.player)) {
+                return;
+            }
+            // Kitin end
             TickThread.ensureTickThread(this.player, "Cannot update player asynchronously");
             if (this.removed) {
                 throw new IllegalStateException("Updating removed player chunk loader");
