From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 26 Dec 2025 16:36:46 +0800
Subject: [PATCH] Network_PlayerThrottling


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 31b028bc07a1100014894fa8e3ab539a66b96ab4..8c544157cd360cc2cb5c019a1bd456b5fd6b0946 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -577,7 +577,15 @@ public final class RegionizedPlayerChunkLoader {
         private double getMaxChunkSendRate() {
             final double configRate = PlatformHooks.get().configPlayerMaxSendRate();
 
-            return configRate <= 0.0 || configRate > (double)MAX_RATE ? (double)MAX_RATE : Math.max(1.0, configRate);
+            double baseRate= configRate <= 0.0 || configRate > (double)MAX_RATE ? (double)MAX_RATE : Math.max(1.0, configRate); //Kitin
+            //Kitin start - Network_PlayerThrottling
+            if (!me.sucixr.kitin.config.KitinConfig.useVanillaChunkSender)
+                return me.sucixr.kitin.network.misc.ChunkSendPolicy
+                        .getInstance()
+                        .limitSendRate(player, baseRate);
+            else
+                return baseRate;
+            //Kitin end
         }
 
         private long getMaxChunkLoads() {
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 6292d9ae1e4b7b586f0806e01b86fc2092f1c153..cd23e14986108847737e25a018d086117eaf4ba1 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1347,6 +1347,14 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 // Paper end - remove allocation of Vec3D here
                 int playerViewDistance = ChunkMap.this.getPlayerViewDistance(player);
                 double d = Math.min(this.getEffectiveRange(), playerViewDistance * 16);
+
+                // Kitin start - Network_PlayerThrottling
+                if (!me.sucixr.kitin.config.KitinConfig.useVanillaChunkSender)
+                    d = me.sucixr.kitin.network.misc.EntitySendPolicy
+                            .getInstance()
+                            .adjustTrackingRange(player, d);
+                // Kitin end
+
                 double d1 = vec3_dx * vec3_dx + vec3_dz * vec3_dz; // Paper
                 double d2 = d * d;
                 // Paper start - Configurable entity tracking range by Y
