From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 26 Dec 2025 16:36:46 +0800
Subject: [PATCH] OPT_Network_PlayerThrottling


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 723b74e5af627acdc8e1436024d723a67c10df1d..5c1b4c402057b3b16ef6100c1dfd8caada2910f6 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -577,7 +577,12 @@ public final class RegionizedPlayerChunkLoader {
         private double getMaxChunkSendRate() {
             final double configRate = PlatformHooks.get().configPlayerMaxSendRate();
 
-            return configRate <= 0.0 || configRate > (double)MAX_RATE ? (double)MAX_RATE : Math.max(1.0, configRate);
+            double baseRate= configRate <= 0.0 || configRate > (double)MAX_RATE ? (double)MAX_RATE : Math.max(1.0, configRate); //Kitin
+            //Kitin start
+            return me.sucixr.kitin.scheduler.policy.ChunkSendPolicy
+                    .getInstance()
+                    .limitSendRate(player, baseRate);
+            //Kitin end
         }
 
         private long getMaxChunkLoads() {
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 6292d9ae1e4b7b586f0806e01b86fc2092f1c153..f63a7d7986e4bdf91a42f3e32dad2c4a09b2b9df 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1347,6 +1347,13 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 // Paper end - remove allocation of Vec3D here
                 int playerViewDistance = ChunkMap.this.getPlayerViewDistance(player);
                 double d = Math.min(this.getEffectiveRange(), playerViewDistance * 16);
+
+                // Kitin start - per-player entity tracking range scaling (network)
+                d = me.sucixr.kitin.scheduler.policy.EntitySendPolicy
+                        .getInstance()
+                        .adjustTrackingRange(player, d);
+                // Kitin end
+
                 double d1 = vec3_dx * vec3_dx + vec3_dz * vec3_dz; // Paper
                 double d2 = d * d;
                 // Paper start - Configurable entity tracking range by Y
