From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 26 Dec 2025 16:36:46 +0800
Subject: [PATCH] OPT_Network_PlayerThrottling


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 723b74e5af627acdc8e1436024d723a67c10df1d..5c1b4c402057b3b16ef6100c1dfd8caada2910f6 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -577,7 +577,12 @@ public final class RegionizedPlayerChunkLoader {
         private double getMaxChunkSendRate() {
             final double configRate = PlatformHooks.get().configPlayerMaxSendRate();
 
-            return configRate <= 0.0 || configRate > (double)MAX_RATE ? (double)MAX_RATE : Math.max(1.0, configRate);
+            double baseRate= configRate <= 0.0 || configRate > (double)MAX_RATE ? (double)MAX_RATE : Math.max(1.0, configRate); //Kitin
+            //Kitin start
+            return me.sucixr.kitin.scheduler.policy.ChunkSendPolicy
+                    .getInstance()
+                    .limitSendRate(player, baseRate);
+            //Kitin end
         }
 
         private long getMaxChunkLoads() {
