From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Sat, 27 Dec 2025 18:42:42 +0800
Subject: [PATCH] 
 OPT_Network_Reduce-redundant-packets-Boat-and-AbstractMinecart


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 3a817c92147bb78c60238f3ddc1644b721aed7a2..4b58170bc7550b211cd6ec5ee7854c8dc2b78777 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -97,7 +97,12 @@ public class ServerEntity {
     public void sendChanges() {
         // Paper start - optimise collisions
         if (((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this.entity).moonrise$isHardColliding()) {
-            this.teleportDelay = 9999;
+            //this.teleportDelay = 9999; Kitin off
+            // Kitin start - throttle forced position sync for vehicles
+            this.teleportDelay =
+                    me.sucixr.kitin.scheduler.policy.EntitySyncPolicy.getInstance()
+                            .adjustTeleportDelayForHardCollision(this.entity, this.tickCount, this.teleportDelay);
+            // Kitin end
         }
         // Paper end - optimise collisions
         this.entity.updateDataBeforeSync();
@@ -175,7 +180,7 @@ public class ServerEntity {
                     || flag5
                     || this.teleportDelay > 400
                     || this.wasRiding
-                    || this.wasOnGround != this.entity.onGround()) {
+                    || (this.wasOnGround != this.entity.onGround()&& !(this.entity instanceof AbstractMinecart || this.entity instanceof net.minecraft.world.entity.item.PrimedTnt))) { //Kitin 优化世吞矿车TNT发包
                     this.wasOnGround = this.entity.onGround();
                     this.teleportDelay = 0;
                     packet = ClientboundEntityPositionSyncPacket.of(this.entity);
