From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 11:17:05 +0800
Subject: [PATCH] Network_Reduce-high-frequency-Entity-packets


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 60af82925d2f7e605bf7340cf01f1a5668f4d725..8a02f4fd34b99bb71f56d0d5e142f3b478955203 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -99,7 +99,12 @@ public class ServerEntity {
     public void sendChanges() {
         // Paper start - optimise collisions
         if (((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this.entity).moonrise$isHardColliding()) {
-            this.teleportDelay = 9999;
+            //this.teleportDelay = 9999; Kitin off
+            // Kitin start - throttle forced position sync for vehicles
+            this.teleportDelay =
+                    me.sucixr.kitin.network.misc.EntitySyncPolicy.getInstance()
+                            .adjustTeleportDelayForHardCollision(this.entity, this.tickCount, this.teleportDelay);
+            // Kitin end
         }
         // Paper end - optimise collisions
         this.entity.updateDataBeforeSync();
@@ -168,6 +173,23 @@ public class ServerEntity {
                 double vec3_dz = currentPosition.z - base.z;
                 boolean positionChanged = (vec3_dx * vec3_dx + vec3_dy * vec3_dy + vec3_dz * vec3_dz) >= 7.62939453125E-6D;
                 // Paper end - reduce allocation of Vec3D here
+
+                // Kitin start - Network_Dont-send-high-speed-entity-packets
+                Vec3 deltaMovement = this.entity.getDeltaMovement();
+                double d = deltaMovement.distanceToSqr(this.lastSentMovement);
+                double currentSpeedSqr = deltaMovement.lengthSqr(); //移动了原版的逻辑位置
+                boolean skipPos = false;
+                double maxSpeedSqr = 65536.0; // viewDistance * 16.0 * 10
+                if (
+                        (this.entity instanceof net.minecraft.world.entity.item.PrimedTnt) && (this.entity.tickCount > 10) //TNT不是下落方块 前10tick不检测防止TNT复制异常
+                ){
+                    boolean isImpact = d > 0.0025;
+                    if (currentSpeedSqr > 0.01 && !isImpact) skipPos = true;
+                }else{
+                    if (!Double.isFinite(currentSpeedSqr) || currentSpeedSqr > maxSpeedSqr) skipPos = true;
+                }
+                // Kitin end
+
                 Packet<ClientGamePacketListener> packet = null;
                 boolean pos = positionChanged || this.tickCount % 60 == 0;
                 boolean sentPosition = false;
@@ -180,22 +202,28 @@ public class ServerEntity {
                     || deltaTooBig
                     || this.teleportDelay > 400
                     || this.wasRiding
-                    || this.wasOnGround != this.entity.onGround()) {
+                    || (this.wasOnGround != this.entity.onGround()&& !(this.entity instanceof AbstractMinecart))){// || this.entity instanceof net.minecraft.world.entity.item.PrimedTnt))) { //Kitin 优化世吞矿车TNT发包
                     this.wasOnGround = this.entity.onGround();
                     this.teleportDelay = 0;
-                    packet = ClientboundEntityPositionSyncPacket.of(this.entity);
-                    sentPosition = true;
+                    if (!skipPos) { // Kitin - Network_Dont-send-high-speed-entity-packets
+                        packet = ClientboundEntityPositionSyncPacket.of(this.entity);
+                        me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "Sync", 60); // Kitin Debug  注入点 A: 强制同步包
+                        sentPosition = true;
+                    } // Kitin
                     sentRotation = true;
                 } else if ((!pos || !shouldSendRotation) && !(this.entity instanceof AbstractArrow)) {
-                    if (pos) {
+                    if (pos && !skipPos) { // Kitin - Network_Dont-send-high-speed-entity-packets
+                        me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "Pos", 15); // Kitin Debug 注入点 B: 纯位移包
                         packet = new ClientboundMoveEntityPacket.Pos(this.entity.getId(), (short)xa, (short)ya, (short)za, this.entity.onGround());
                         sentPosition = true;
                     } else if (shouldSendRotation) {
+                        me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "Rot", 10); // Kitin Debug 注入点 C: 纯旋转包
                         packet = new ClientboundMoveEntityPacket.Rot(this.entity.getId(), yRotn, xRotn, this.entity.onGround());
                         sentRotation = true;
                     }
                 } else {
                     packet = new ClientboundMoveEntityPacket.PosRot(this.entity.getId(), (short)xa, (short)ya, (short)za, yRotn, xRotn, this.entity.onGround());
+                    me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "PosRot", 20); // Kitin Debug 注入点 D: 位移+旋转包
                     sentPosition = true;
                     sentRotation = true;
                 }
@@ -216,7 +244,21 @@ public class ServerEntity {
                                     )
                                 );
                         } else {
-                            this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                            // Kitin start - Network_Dont-send-high-speed-entity-packets
+                            boolean skipMotion = false;
+                            if (this.entity instanceof net.minecraft.world.entity.item.ItemEntity
+                                    || this.entity instanceof net.minecraft.world.entity.ExperienceOrb
+                                    || this.entity instanceof net.minecraft.world.entity.item.PrimedTnt) {
+                                boolean isImpact = d > 0.0025;
+                                if (currentSpeedSqr > 0.01 && !isImpact) skipMotion = true;
+                            }else{
+                                if (!Double.isFinite(currentSpeedSqr) || currentSpeedSqr > maxSpeedSqr) skipMotion = true;
+                            }
+                            if (!skipMotion) {
+                                this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                            }
+                            // Kitin end
+                            if (!skipMotion) me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "Motion", 26); // Kitin Debug 注入点 E: 速度包
                         }
                     }
                 }
