From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Thu, 1 Jan 2026 05:17:26 +0800
Subject: [PATCH] FoliaFix_Ender-Pearl-Chunk-Loading


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 9b0e4e0d0555dcaf1c4323dec7057235588877a0..fda4387cd19d23c1cef796a39d76f4f51cf8ca28 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -757,6 +757,10 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     private void saveEnderPearls(ValueOutput output) {
+        // Kitin start - fix Pearl Chunk Loading
+        if (me.sucixr.kitin.config.KitinConfig.pearlFixEnabled)me.sucixr.kitin.fixes.foliafix.pearl.PearlChunkLoading.saveEnderPearls(this, output);
+        else {
+        // Kitin end
         if (!this.enderPearls.isEmpty()) {
             ValueOutput.ValueOutputList valueOutputList = output.childrenList("ender_pearls");
 
@@ -770,7 +774,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     valueOutput.store("ender_pearl_dimension", Level.RESOURCE_KEY_CODEC, thrownEnderpearl.level().dimension());
                 }
             }
-        }
+        }} //Kitin
     }
 
     public void loadAndSpawnEnderPearls(ValueInput input) {
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 7d8c528a76e6a95d27368bc459d654c6ef66dd8f..03a45923355161f879ad614962c5ec75a07a04ce 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -174,9 +174,21 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                 if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this)) {
                     throw new IllegalStateException("Must be on tick thread for ticking entity: " + this);
                 }
-                Entity entity = this.getOwnerRaw();
+                //Kitin start - fix Pearl Chunk Loading
+                Entity entity = null;
+                if (this.owner != null) {
+                    UUID ownerUUID = this.owner.getUUID();
+                    entity = serverLevel.getServer().getPlayerList().getPlayer(ownerUUID);
+                }
+                //Kitin end
                 if (entity != null) {
                     attemptTeleport(entity, (ServerLevel)this.level(), this.position());
+                    //Kitin start - fix Pearl Chunk Loading
+                    var pearls = me.sucixr.kitin.fixes.foliafix.pearl.PearlSnapshot.PEARL_DATA_MAP.get(this.owner.getUUID());
+                    if (pearls != null) {
+                        pearls.remove(this.getUUID());
+                    }
+                    //Kitin end
                 }
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
                 return;
@@ -253,11 +265,21 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
     @Override
     public void tick() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            // Kitin start - Pearl Chunk Loading
+            if (me.sucixr.kitin.config.KitinConfig.pearlFixEnabled) {
+                final java.util.UUID ownerId = this.owner != null ? this.owner.getUUID() : null;
+                if (me.sucixr.kitin.fixes.foliafix.pearl.PearlSnapshot.tickPearl(this, ownerId)) {
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                    return;
+                }
+            }
+            // Kitin end
             int var7 = SectionPos.blockToSectionCoord(this.position().x());
             int sectionPosZ = SectionPos.blockToSectionCoord(this.position().z());
-            Entity entity = this.owner != null ? findOwnerIncludingDeadPlayer(serverLevel, this.owner.getUUID()) : null;
+            Entity entity = this.owner != null ? serverLevel.getServer().getPlayerList().getPlayer(this.owner.getUUID()) : null; //Kitin fix Pearl Chunk Loading
             if (entity instanceof ServerPlayer serverPlayer
                 && !entity.isAlive()
+                && serverPlayer.getHealth() <= 0.0F //Kitin fix - Pearl Chunk Loading
                 && !serverPlayer.wonGame
                 && serverPlayer.level().getGameRules().get(GameRules.ENDER_PEARLS_VANISH_ON_DEATH)) {
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
@@ -272,9 +294,17 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                             || var7 != SectionPos.blockToSectionCoord(blockPos.getX())
                             || sectionPosZ != SectionPos.blockToSectionCoord(blockPos.getZ())
                     )
-                    && entity instanceof ServerPlayer serverPlayer1) {
-                    this.ticketTimer = serverPlayer1.registerAndUpdateEnderPearlTicket(this);
-                }
+                    //Kitin start - fix Pearl Chunk Loading
+                )
+                    //&& entity instanceof ServerPlayer serverPlayer1) {
+                    if (me.sucixr.kitin.config.KitinConfig.pearlFixEnabled && !this.isRemoved()){
+                        serverLevel.resetEmptyTime(); // 防止维度卸载
+                        net.minecraft.server.level.ServerPlayer.placeEnderPearlTicket(serverLevel, this.chunkPosition());
+                        this.ticketTimer = net.minecraft.server.level.TicketType.ENDER_PEARL.timeout();
+                    } else if (entity instanceof ServerPlayer serverPlayer1) {
+                        this.ticketTimer = serverPlayer1.registerAndUpdateEnderPearlTicket(this);
+                    }
+                //Kitin end
             }
         } else {
             super.tick();
@@ -322,6 +352,12 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
     public void onRemoval(Entity.RemovalReason reason) {
         if (reason != Entity.RemovalReason.UNLOADED_WITH_PLAYER) {
             this.deregisterFromCurrentOwner();
+            //Kitin start - fix Pearl Chunk Loading
+            var pearls = me.sucixr.kitin.fixes.foliafix.pearl.PearlSnapshot.PEARL_DATA_MAP.get(this.owner.getUUID());
+            if (pearls != null) {
+                pearls.remove(this.getUUID());
+            }
+            //Kitin end
         }
 
         super.onRemoval(reason);
