From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 11:26:44 +0800
Subject: [PATCH] FoliaFix_Ender-Pearl-Chunk-Loading


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index cc9b535f0ea457eecf019375403f18bf967ba900..dc1502d6017a25a8185b74fe7f82459168883a89 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -768,6 +768,10 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     private void saveEnderPearls(final ValueOutput playerOutput) {
+        // Kitin start - fix Pearl Chunk Loading
+        if (me.sucixr.kitin.config.KitinConfig.pearlFixEnabled) me.sucixr.kitin.fixes.foliafix.pearl.PearlChunkLoading.saveEnderPearls(this, playerOutput);
+        else {
+        // Kitin end
         if (!this.enderPearls.isEmpty()) {
             ValueOutput.ValueOutputList pearlsOutput = playerOutput.childrenList("ender_pearls");
 
@@ -781,7 +785,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     pearlTag.store("ender_pearl_dimension", Level.RESOURCE_KEY_CODEC, enderPearl.level().dimension());
                 }
             }
-        }
+        }} //Kitin
     }
 
     public void loadAndSpawnEnderPearls(final ValueInput playerInput) {
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index d0dba5cc351903fb762e231a95f302479794185f..045d5538bde5d71eaf959ee67c96b00862875207 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -176,9 +176,17 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                 if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this)) {
                     throw new IllegalStateException("Must be on tick thread for ticking entity: " + this);
                 }
-                Entity entity = this.getOwnerRaw();
+                //Kitin start - fix Pearl Chunk Loading
+                Entity entity = me.sucixr.kitin.fixes.util.CrossRegionEntityUtil.getGlobalPlayer(level.getServer(), this.owner != null ? this.owner.getUUID() : null);
+                //Kitin end
                 if (entity != null) {
                     attemptTeleport(entity, (ServerLevel)this.level(), this.position());
+                    //Kitin start - fix Pearl Chunk Loading
+                    var pearls = me.sucixr.kitin.fixes.foliafix.pearl.PearlSnapshot.PEARL_DATA_MAP.get(this.owner.getUUID());
+                    if (pearls != null) {
+                        pearls.remove(this.getUUID());
+                    }
+                    //Kitin end
                 }
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
                 return;
@@ -260,11 +268,21 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
     @Override
     public void tick() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            // Kitin start - Pearl Chunk Loading
+            if (me.sucixr.kitin.config.KitinConfig.pearlFixEnabled) {
+                final java.util.UUID ownerId = this.owner != null ? this.owner.getUUID() : null;
+                if (me.sucixr.kitin.fixes.foliafix.pearl.PearlSnapshot.tickPearl(this, ownerId)) {
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                    return;
+                }
+            }
+            // Kitin end
             int var7 = SectionPos.blockToSectionCoord(this.position().x());
             int previousChunkZ = SectionPos.blockToSectionCoord(this.position().z());
-            Entity owner = this.owner != null ? findOwnerIncludingDeadPlayer(serverLevel, this.owner.getUUID()) : null;
+            Entity owner = me.sucixr.kitin.fixes.util.CrossRegionEntityUtil.getGlobalPlayer(serverLevel.getServer(), this.owner != null ? this.owner.getUUID() : null); //Kitin fix Pearl Chunk Loading
             if (owner instanceof ServerPlayer serverPlayer
                 && !owner.isAlive()
+                && serverPlayer.getHealth() <= 0.0F //Kitin fix - Pearl Chunk Loading
                 && !serverPlayer.wonGame
                 && serverPlayer.level().getGameRules().get(GameRules.ENDER_PEARLS_VANISH_ON_DEATH)) {
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
@@ -279,9 +297,17 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                             || var7 != SectionPos.blockToSectionCoord(currentPos.getX())
                             || previousChunkZ != SectionPos.blockToSectionCoord(currentPos.getZ())
                     )
-                    && owner instanceof ServerPlayer serverPlayer) {
-                    this.ticketTimer = serverPlayer.registerAndUpdateEnderPearlTicket(this);
-                }
+                    //Kitin start - fix Pearl Chunk Loading
+                )
+                    //&& owner instanceof ServerPlayer serverPlayer) {
+                    if (me.sucixr.kitin.config.KitinConfig.pearlFixEnabled && !this.isRemoved()){
+                        serverLevel.resetEmptyTime(); // 防止维度卸载
+                        net.minecraft.server.level.ServerPlayer.placeEnderPearlTicket(serverLevel, this.chunkPosition());
+                        this.ticketTimer = net.minecraft.server.level.TicketType.ENDER_PEARL.timeout();
+                    } else if (owner instanceof ServerPlayer serverPlayer1) {
+                        this.ticketTimer = serverPlayer1.registerAndUpdateEnderPearlTicket(this);
+                    }
+                //Kitin end
             }
         } else {
             super.tick();
@@ -329,6 +355,12 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
     public void onRemoval(final Entity.RemovalReason reason) {
         if (reason != Entity.RemovalReason.UNLOADED_WITH_PLAYER) {
             this.deregisterFromCurrentOwner();
+            //Kitin start - fix Pearl Chunk Loading
+            var pearls = me.sucixr.kitin.fixes.foliafix.pearl.PearlSnapshot.PEARL_DATA_MAP.get(this.owner.getUUID());
+            if (pearls != null) {
+                pearls.remove(this.getUUID());
+            }
+            //Kitin end
         }
 
         super.onRemoval(reason);
