From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Thu, 1 Jan 2026 05:17:26 +0800
Subject: [PATCH] FoliaFix_Pearl-Chunk-Loading


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 9b70056682c9b122dbe542474d31bf16d3fe2ae2..58f48cb8c7f9b9bf4fec4eee74d25bd5ef7db61c 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -757,20 +757,21 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     private void saveEnderPearls(ValueOutput output) {
-        if (!this.enderPearls.isEmpty()) {
-            ValueOutput.ValueOutputList valueOutputList = output.childrenList("ender_pearls");
-
-            for (ThrownEnderpearl thrownEnderpearl : this.enderPearls) {
-                if (thrownEnderpearl.level().paperConfig().misc.legacyEnderPearlBehavior) continue; // Paper - Allow using old ender pearl behavior
-                if (thrownEnderpearl.isRemoved()) {
-                    LOGGER.warn("Trying to save removed ender pearl, skipping");
-                } else {
-                    ValueOutput valueOutput = valueOutputList.addChild();
-                    thrownEnderpearl.save(valueOutput);
-                    valueOutput.store("ender_pearl_dimension", Level.RESOURCE_KEY_CODEC, thrownEnderpearl.level().dimension());
-                }
-            }
-        }
+        me.sucixr.kitin.fixes.foliafix.PearlChunkLoading.saveEnderPearls(this, output);
+//        if (!this.enderPearls.isEmpty()) {
+//            ValueOutput.ValueOutputList valueOutputList = output.childrenList("ender_pearls");
+//
+//            for (ThrownEnderpearl thrownEnderpearl : this.enderPearls) {
+//                if (thrownEnderpearl.level().paperConfig().misc.legacyEnderPearlBehavior) continue; // Paper - Allow using old ender pearl behavior
+//                if (thrownEnderpearl.isRemoved()) {
+//                    LOGGER.warn("Trying to save removed ender pearl, skipping");
+//                } else {
+//                    ValueOutput valueOutput = valueOutputList.addChild();
+//                    thrownEnderpearl.save(valueOutput);
+//                    valueOutput.store("ender_pearl_dimension", Level.RESOURCE_KEY_CODEC, thrownEnderpearl.level().dimension());
+//                }
+//            }
+//        } Kitin off
     }
 
     public void loadAndSpawnEnderPearls(ValueInput input) {
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 7d8c528a76e6a95d27368bc459d654c6ef66dd8f..5e77b22ad395c219f8073d1ef134d58d145fdc50 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -174,9 +174,20 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                 if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this)) {
                     throw new IllegalStateException("Must be on tick thread for ticking entity: " + this);
                 }
-                Entity entity = this.getOwnerRaw();
+                //Kitin start - fix Pearl Chunk Loading
+                Entity entity = null;
+                if (this.owner != null) {
+                    UUID ownerUUID = this.owner.getUUID();
+                    entity = serverLevel.getServer().getPlayerList().getPlayer(ownerUUID);
+                }
+                //Kitin end
                 if (entity != null) {
                     attemptTeleport(entity, (ServerLevel)this.level(), this.position());
+                    //Kitin start - fix Pearl Chunk Loading
+                    if (this.owner != null) {
+                        me.sucixr.kitin.scheduler.data.PearlSnapshot.PEARL_DATA_MAP.remove(this.owner.getUUID());
+                    }
+                    //Kitin end
                 }
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
                 return;
@@ -253,11 +264,19 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
     @Override
     public void tick() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            // Kitin start - Pearl Chunk Loading
+            final java.util.UUID ownerId = this.owner != null ? this.owner.getUUID() : null;
+            if (me.sucixr.kitin.scheduler.data.PearlSnapshot.tickPearl(this,ownerId)) {
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                return;
+            }
+            // Kitin end
             int var7 = SectionPos.blockToSectionCoord(this.position().x());
             int sectionPosZ = SectionPos.blockToSectionCoord(this.position().z());
-            Entity entity = this.owner != null ? findOwnerIncludingDeadPlayer(serverLevel, this.owner.getUUID()) : null;
+            Entity entity = this.owner != null ? serverLevel.getServer().getPlayerList().getPlayer(this.owner.getUUID()) : null; //Kitin fix Pearl Chunk Loading
             if (entity instanceof ServerPlayer serverPlayer
                 && !entity.isAlive()
+                && serverPlayer.getHealth() <= 0.0F //Kitin fix - Pearl Chunk Loading
                 && !serverPlayer.wonGame
                 && serverPlayer.level().getGameRules().get(GameRules.ENDER_PEARLS_VANISH_ON_DEATH)) {
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
@@ -272,9 +291,15 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                             || var7 != SectionPos.blockToSectionCoord(blockPos.getX())
                             || sectionPosZ != SectionPos.blockToSectionCoord(blockPos.getZ())
                     )
-                    && entity instanceof ServerPlayer serverPlayer1) {
-                    this.ticketTimer = serverPlayer1.registerAndUpdateEnderPearlTicket(this);
-                }
+                    //Kitin start - fix Pearl Chunk Loading
+                )
+                    //&& entity instanceof ServerPlayer serverPlayer1) {
+                    if (!this.isRemoved()){
+                        serverLevel.resetEmptyTime(); // 防止维度卸载
+                        net.minecraft.server.level.ServerPlayer.placeEnderPearlTicket(serverLevel, this.chunkPosition());
+                        this.ticketTimer = net.minecraft.server.level.TicketType.ENDER_PEARL.timeout();
+                    }
+                //Kitin end
             }
         } else {
             super.tick();
@@ -322,6 +347,11 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
     public void onRemoval(Entity.RemovalReason reason) {
         if (reason != Entity.RemovalReason.UNLOADED_WITH_PLAYER) {
             this.deregisterFromCurrentOwner();
+            //Kitin start - fix Pearl Chunk Loading
+            if (this.owner != null) {
+                me.sucixr.kitin.scheduler.data.PearlSnapshot.PEARL_DATA_MAP.remove(this.owner.getUUID());
+            }
+            //Kitin end
         }
 
         super.onRemoval(reason);
