From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 11:33:46 +0800
Subject: [PATCH] SpigotFix_Primed-and-fix-TNT


diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 42aff39d537aa6b138bc4a76304d49f423212acd..0a1e1716457445048dad5bf39fb0c1b824d5e897 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -101,7 +101,8 @@ public class PrimedTnt extends Entity implements TraceableEntity {
 
     @Override
     public void tick() {
-        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().getCurrentWorldData().currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { return; } // Spigot // Folia - region threading
+        boolean isOverloaded = false;//Kitin - TNT OPT
+        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().getCurrentWorldData().currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { if (me.sucixr.kitin.config.KitinConfig.disableMaxTntPerTickAndOptimize) isOverloaded = true; else return;} // Spigot // Folia - region threading //Kitin
         //this.handlePortal(); // Folia - region threading
         this.applyGravity();
         this.move(MoverType.SELF, this.getDeltaMovement());
@@ -129,7 +130,8 @@ public class PrimedTnt extends Entity implements TraceableEntity {
             // CraftBukkit end
         } else {
             this.updateFluidInteraction();
-            if (this.level().isClientSide()) {
+            //if (!isOverloaded) {this.updateInWaterStateAndDoFluidPushing();} //Kitin TNT水流推力计算去除,暂时注释，部分复制机需要水
+            if (!me.sucixr.kitin.config.KitinConfig.disableMaxTntPerTickAndOptimize || !isOverloaded && this.level().isClientSide()) {//Kitin TNT爆炸粒子去除
                 this.level().addParticle(ParticleTypes.SMOKE, this.getX(), this.getY() + 0.5, this.getZ(), 0.0, 0.0, 0.0);
             }
         }
