From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 11:35:11 +0800
Subject: [PATCH] Safety_LazyChunkBarrier


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 223793bea54a483fefc66caa7e07b29ab351c339..04394ff59c7669ea600d907a39fe5197116a4783 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -61,6 +61,14 @@ public class ItemEntity extends Entity implements TraceableEntity {
         this.setYRot(this.random.nextFloat() * 360.0F);
     }
 
+    //Kitin start - LazyChunkBarrier
+    @Override
+    public void setPos(double x, double y, double z) {
+        super.setPos(x, y, z);
+        me.sucixr.kitin.safety.chunkbarrier.LazyChunkBarrierPolicy.onCheckMovement(this);
+    }
+    //Kitin end
+
     public ItemEntity(final Level level, final double x, final double y, final double z, final ItemStack itemStack) {
         this(EntityTypes.ITEM, level);
         this.setPos(x, y, z);
