From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 11:35:11 +0800
Subject: [PATCH] Safety_LazyChunkBarrier


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 4f39b96a4af8865af6a968d64643712840359851..7ed1d6c5fe341185b713ff786834c7935d2381b8 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -60,6 +60,14 @@ public class ItemEntity extends Entity implements TraceableEntity {
         this.setYRot(this.random.nextFloat() * 360.0F);
     }
 
+    //Kitin start - LazyChunkBarrier
+    @Override
+    public void setPos(double x, double y, double z) {
+        super.setPos(x, y, z);
+        me.sucixr.kitin.safety.chunkbarrier.LazyChunkBarrierPolicy.onCheckMovement(this);
+    }
+    //Kitin end
+
     public ItemEntity(final Level level, final double x, final double y, final double z, final ItemStack itemStack) {
         this(EntityType.ITEM, level);
         this.setPos(x, y, z);
