From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Sun, 11 Jan 2026 22:16:17 +0800
Subject: [PATCH] SpigotFIX_OPT_Primed-and-fix-TNT


diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index fd1db18764be58bc5ed5983b98e669c76f7ff427..a5a7cc2763b3fdab9391a81d10a0feb883cd735e 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -99,11 +99,12 @@ public class PrimedTnt extends Entity implements TraceableEntity {
 
     @Override
     public void tick() {
-        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().getCurrentWorldData().currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { return; } // Spigot // Folia - region threading
+        boolean isOverloaded = false;//Kitin - TNT OPT
+        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().getCurrentWorldData().currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { isOverloaded = true; } // Spigot // Folia - region threading //Kitin
         //this.handlePortal(); // Folia - region threading
         this.applyGravity();
         this.move(MoverType.SELF, this.getDeltaMovement());
-        this.applyEffectsFromBlocks();
+        if (!isOverloaded) {this.applyEffectsFromBlocks();}//Kitin TNT不计算环境，水流岩浆，气泡柱等
         // Paper start - Configurable TNT height nerf
         if (this.level().paperConfig().fixes.tntEntityHeightNerf.test(v -> this.getY() > v)) {
             this.discard(EntityRemoveEvent.Cause.OUT_OF_WORLD);
@@ -114,12 +115,11 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         if (this.onGround()) {
             this.setDeltaMovement(this.getDeltaMovement().multiply(0.7, -0.5, 0.7));
         }
-
         int i = this.getFuse() - 1;
         this.setFuse(i);
         if (i <= 0) {
             // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
-            //this.discard();
+            this.discard();//Kitin re
             if (!this.level().isClientSide()) {
                 this.explode();
             }
@@ -127,7 +127,8 @@ public class PrimedTnt extends Entity implements TraceableEntity {
             // CraftBukkit end
         } else {
             this.updateInWaterStateAndDoFluidPushing();
-            if (this.level().isClientSide()) {
+            //if (!isOverloaded) {this.updateInWaterStateAndDoFluidPushing();} //Kitin TNT水流推力计算去除,暂时注释，部分复制机需要水
+            if (!isOverloaded && this.level().isClientSide()) {//Kitin TNT爆炸粒子去除
                 this.level().addParticle(ParticleTypes.SMOKE, this.getX(), this.getY() + 0.5, this.getZ(), 0.0, 0.0, 0.0);
             }
         }
