From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:00:00 +0800
Subject: [PATCH] Perf_VillagerSmartHibernation


diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index f649c46cc9a34e646f8da3244866c6e011d30aa3..eff2023fd6470e7017418c4c1b3661f4375189e9 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -171,6 +171,7 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
             return activities;
         }
     );
+    public net.minecraft.world.level.block.state.BlockState[] kitin$prisonSnapshot = null; // Kitin - VillagerSmartHibernation
     public static final Map<MemoryModuleType<GlobalPos>, BiPredicate<Villager, Holder<PoiType>>> POI_MEMORIES = ImmutableMap.of(
         MemoryModuleType.HOME,
         (villager, poiType) -> poiType.is(PoiTypes.HOME),
@@ -261,6 +262,7 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
     }
     protected void customServerAiStep(ServerLevel level, final boolean inactive) {
         // Paper end - EAR 2
+        if (me.sucixr.kitin.performance.entity.villager.VillagerSmartHibernation.checkFreezeTick(this, level)) return; // Kitin - VillagerSmartHibernation
         ProfilerFiller profiler = Profiler.get();
         profiler.push("villagerBrain");
         if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
@@ -300,6 +302,7 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
 
         if (inactive) return; // Paper - EAR 2
         super.customServerAiStep(level);
+        me.sucixr.kitin.performance.entity.villager.VillagerSmartHibernation.attemptCapture(this, level); // Kitin - VillagerSmartHibernation
     }
 
     @Override
