From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:03:10 +0800
Subject: [PATCH] Network_Particle-throttling-and-packing


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 4924915ec3adf4d68ca537518fdffa52ec926bf7..374a6a22871e9a126f10753a039e8f8fa2afc730 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -27,6 +27,7 @@ import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import me.sucixr.kitin.network.particle.ParticleOptimizer;
 import net.minecraft.ChatFormatting;
 import net.minecraft.advancements.AdvancementHolder;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -336,6 +337,17 @@ public class ServerGamePacketListenerImpl
     public final Long disconnectTicketId = Long.valueOf(DISCONNECT_TICKET_ID_GENERATOR.getAndIncrement());
     // Folia end - region threading
 
+    // Kitin start - ParticleOptimizer
+    private final ParticleOptimizer kitin$particleOptimizer;
+    @Override
+    public void send(Packet<?> packet) {
+        if (this.kitin$particleOptimizer.handleSend(packet, this)) {
+            return;
+        }
+        super.send(packet);
+    }
+    //Kitin end
+
     public ServerGamePacketListenerImpl(final MinecraftServer server, final Connection connection, final ServerPlayer player, final CommonListenerCookie cookie) {
         super(server, connection, cookie);
         this.restartClientLoadTimerAfterRespawn();
@@ -347,6 +359,7 @@ public class ServerGamePacketListenerImpl
         this.chatMessageChain = new FutureChain(server.chatExecutor); // CraftBukkit - async chat
         this.tickEndEvent = new io.papermc.paper.event.packet.ClientTickEndEvent(player.getBukkitEntity()); // Paper - add client tick end event
         this.playerGameConnection =  new io.papermc.paper.connection.PaperPlayerGameConnection(this); // Paper
+        this.kitin$particleOptimizer = new ParticleOptimizer(player); // Kitin - ParticleOptimizer
     }
 
     // Paper start - configuration phase API
@@ -364,6 +377,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void tick() {
         // Folia start - region threading
+        this.kitin$particleOptimizer.tick(this); // Kitin - ParticleOptimizer
         this.keepConnectionAlive();
         if (this.player.wonGame) {
             return;
