From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Thu, 15 Jan 2026 17:05:52 +0800
Subject: [PATCH] FoliaFix_FunctionCommand


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index b83e83cb0c202a6f0ac91397e1593e1042a666d3..b5400efbc0537536810c12b4d9184a1450dc8c12 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -213,7 +213,7 @@ public class Commands {
         FillCommand.register(this.dispatcher, context);
         FillBiomeCommand.register(this.dispatcher, context);
         ForceLoadCommand.register(this.dispatcher);
-        //FunctionCommand.register(this.dispatcher); // Folia - region threading - TODO
+        FunctionCommand.register(this.dispatcher); // Folia - region threading - TODO //Kitin re
         GameModeCommand.register(this.dispatcher);
         GameRuleCommand.register(this.dispatcher, context);
         GiveCommand.register(this.dispatcher, context);
