From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Wed, 21 Jan 2026 22:28:33 +0800
Subject: [PATCH] Perf_dropper-container-transfer


diff --git a/net/minecraft/world/level/block/DropperBlock.java b/net/minecraft/world/level/block/DropperBlock.java
index 0d950a6b32524d0d982d7856fc4791bf3670938e..ef9c58e3eae1ce376ac8c36eb0eb8391c67c836d 100644
--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -65,6 +65,36 @@ public class DropperBlock extends DispenserBlock {
                         if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(level, pos, item, randomSlot)) return; // Paper - Add BlockPreDispenseEvent
                         itemStack = DISPENSE_BEHAVIOUR.dispense(blockSource, item);
                     } else {
+                        // Kitin start
+                        if (me.sucixr.kitin.config.KitinConfig.optimizeDropper){
+                            // 1. 预检
+                            int freeSlot = -1;
+                            for (int i = 0; i < containerAt.getContainerSize(); i++) {
+                                ItemStack target = containerAt.getItem(i);
+                                if (target.isEmpty() || (ItemStack.isSameItemSameComponents(target, item) && target.getCount() < target.getMaxStackSize())) {
+                                    freeSlot = i;
+                                    break;
+                                }
+                            }
+                            // 2. 熔断
+                            if (freeSlot == -1) {
+                                return;
+                            }
+                            // 3. 零拷贝移动
+                            ItemStack toMove = item.split(1);
+                            ItemStack targetAtSlot = containerAt.getItem(freeSlot);
+
+                            if (targetAtSlot.isEmpty()) {
+                                containerAt.setItem(freeSlot, toMove);
+                            } else {
+                                targetAtSlot.grow(1);
+                            }
+
+                            dispenserBlockEntity.setItem(randomSlot, item);
+                            return;
+                        }
+                        // Kitin end
+
                         // CraftBukkit start - Fire event when pushing items into other inventories
                         org.bukkit.craftbukkit.inventory.CraftItemStack oitemstack = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(item.copyWithCount(1));
 
