From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Wed, 21 Jan 2026 15:38:54 +0800
Subject: [PATCH] Other_add-Debug


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index e6a2be26c3e9ad0c9ece6166194511cc636c1f38..c1fd892fa0b8c01a58442488622a56550ae69575 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -168,23 +168,6 @@ public class ServerEntity {
                 double vec3_dz = vec3.z - base.z;
                 boolean flag1 = (vec3_dx * vec3_dx + vec3_dy * vec3_dy + vec3_dz * vec3_dz) >= 7.62939453125E-6D;
                 // Paper end - reduce allocation of Vec3D here
-
-                // Kitin start - Network_Dont-send-high-speed-entity-packets
-                Vec3 deltaMovement = this.entity.getDeltaMovement();
-                double d = deltaMovement.distanceToSqr(this.lastSentMovement);
-                double currentSpeedSqr = deltaMovement.lengthSqr(); //移动了原版的逻辑位置
-                boolean skipPos = false;
-                double maxSpeedSqr = 65536.0; // viewDistance * 16.0 * 10
-                if (
-                        (this.entity instanceof net.minecraft.world.entity.item.PrimedTnt) && (this.entity.tickCount > 2) //TNT不是下落方块 前2tick不检测防止世吞漏尿
-                ){
-                    boolean isImpact = d > 0.0025;
-                    if (currentSpeedSqr > 0.01 && !isImpact) skipPos = true;
-                }else{
-                    if (!Double.isFinite(currentSpeedSqr) || currentSpeedSqr > maxSpeedSqr) skipPos = true;
-                }
-                // Kitin end
-
                 Packet<ClientGamePacketListener> packet = null;
                 boolean flag2 = flag1 || this.tickCount % 60 == 0;
                 boolean flag3 = false;
@@ -200,12 +183,12 @@ public class ServerEntity {
                     || (this.wasOnGround != this.entity.onGround()&& !(this.entity instanceof AbstractMinecart || this.entity instanceof net.minecraft.world.entity.item.PrimedTnt))) { //Kitin 优化世吞矿车TNT发包
                     this.wasOnGround = this.entity.onGround();
                     this.teleportDelay = 0;
-                    if (!skipPos) packet = ClientboundEntityPositionSyncPacket.of(this.entity); // Kitin - Network_Dont-send-high-speed-entity-packets
+                    packet = ClientboundEntityPositionSyncPacket.of(this.entity);
                     flag3 = true;
                     flag4 = true;
-                    me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "Sync", 60); // Kitin Debug  注入点 A: 强制同步包
+                    if (!skipPos) me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "Sync", 60); // Kitin Debug  注入点 A: 强制同步包
                 } else if ((!flag2 || !flag) && !(this.entity instanceof AbstractArrow)) {
-                    if (flag2 && !skipPos) { // Kitin - Network_Dont-send-high-speed-entity-packets
+                    if (flag2) {
                         me.sucixr.kitin.other.debug.EntityPacketDebug.record(this.entity, "Pos", 15); // Kitin Debug 注入点 B: 纯位移包
                         packet = new ClientboundMoveEntityPacket.Pos(this.entity.getId(), (short)l, (short)l1, (short)l2, this.entity.onGround());
                         flag3 = true;
@@ -222,6 +205,9 @@ public class ServerEntity {
                 }
 
                 if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying()) {
+                    Vec3 deltaMovement = this.entity.getDeltaMovement();
+                    double d = deltaMovement.distanceToSqr(this.lastSentMovement);
+                    double currentSpeedSqr = deltaMovement.lengthSqr(); // Kitin Network_Dont-send-high-speed-entity-packets
                     if (d > 1.0E-7 || d > 0.0 && currentSpeedSqr == 0.0) { // Kitin Network_Dont-send-high-speed-entity-packets
                         this.lastSentMovement = deltaMovement;
                         if (this.entity instanceof AbstractHurtingProjectile abstractHurtingProjectile) {
@@ -237,9 +223,9 @@ public class ServerEntity {
                         } else {
                             // Kitin start - Network_Dont-send-high-speed-entity-packets
                             boolean skipMotion = false;
+                            double maxSpeedSqr = 65536.0; // viewDistance * 16.0 * 10
                             if (this.entity instanceof net.minecraft.world.entity.item.ItemEntity
-                                    || this.entity instanceof net.minecraft.world.entity.ExperienceOrb
-                                    || this.entity instanceof net.minecraft.world.entity.item.PrimedTnt) {
+                                    || this.entity instanceof net.minecraft.world.entity.ExperienceOrb) {
                                 boolean isImpact = d > 0.0025;
                                 if (currentSpeedSqr > 0.01 && !isImpact) skipMotion = true;
                             }else{
