From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Wed, 21 Jan 2026 15:37:36 +0800
Subject: [PATCH] 
 Network_Dont-send-high-speed-ItemEntity-and-ExperienceOrb-packets


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 4eeaee3871f1647de351032d32329c3fead2bd56..5524dfaa58f32d30a630c44bd3d101856b3e327b 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -216,7 +216,20 @@ public class ServerEntity {
                                     )
                                 );
                         } else {
-                            this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                            // Kitin Network_Dont-send-high-speed-entity-packets
+                            boolean skipMotion = false;
+                            if (this.entity instanceof net.minecraft.world.entity.item.ItemEntity
+                                    || this.entity instanceof net.minecraft.world.entity.ExperienceOrb) {
+                                double currentSpeedSqr = this.lastSentMovement.lengthSqr();
+                                boolean isImpact = d > 0.0025;
+                                if (currentSpeedSqr > 0.01 && !isImpact) {
+                                    skipMotion = true;
+                                }
+                            }
+                            if (!skipMotion) {
+                                this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                            }
+                            // Kitin end
                         }
                     }
                 }
