From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Wed, 21 Jan 2026 22:28:33 +0800
Subject: [PATCH] Perf_dropper-container-transfer


diff --git a/net/minecraft/world/level/block/DropperBlock.java b/net/minecraft/world/level/block/DropperBlock.java
index 0d950a6b32524d0d982d7856fc4791bf3670938e..bd89eba00e992e8bd4d9770ca683d4e776c48df4 100644
--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -65,31 +65,59 @@ public class DropperBlock extends DispenserBlock {
                         if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(level, pos, item, randomSlot)) return; // Paper - Add BlockPreDispenseEvent
                         itemStack = DISPENSE_BEHAVIOUR.dispense(blockSource, item);
                     } else {
-                        // CraftBukkit start - Fire event when pushing items into other inventories
-                        org.bukkit.craftbukkit.inventory.CraftItemStack oitemstack = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(item.copyWithCount(1));
-
-                        org.bukkit.inventory.Inventory destinationInventory;
-                        // Have to special case large chests as they work oddly
-                        if (containerAt instanceof net.minecraft.world.CompoundContainer compoundContainer) {
-                            destinationInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest(compoundContainer);
-                        } else {
-                            destinationInventory = containerAt.getOwner().getInventory();
+                        // Kitin start
+                        // 1. 预检
+                        int freeSlot = -1;
+                        for (int i = 0; i < containerAt.getContainerSize(); i++) {
+                            ItemStack target = containerAt.getItem(i);
+                            if (target.isEmpty() || (ItemStack.isSameItemSameComponents(target, item) && target.getCount() < target.getMaxStackSize())) {
+                                freeSlot = i;
+                                break;
+                            }
                         }
-
-                        org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(dispenserBlockEntity.getOwner().getInventory(), oitemstack, destinationInventory, true);
-                        level.getCraftServer().getPluginManager().callEvent(event);
-                        if (event.isCancelled()) {
+                        // 2. 熔断
+                        if (freeSlot == -1) {
                             return;
                         }
-                        itemStack = HopperBlockEntity.addItem(dispenserBlockEntity, containerAt, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem()), direction.getOpposite());
-                        if (event.getItem().equals(oitemstack) && itemStack.isEmpty()) {
-                            // CraftBukkit end
+                        // 3. 零拷贝移动
+                        ItemStack toMove = item.split(1);
+                        ItemStack targetAtSlot = containerAt.getItem(freeSlot);
 
-                            itemStack = item.copy();
-                            itemStack.shrink(1);
+                        if (targetAtSlot.isEmpty()) {
+                            containerAt.setItem(freeSlot, toMove);
                         } else {
-                            itemStack = item.copy();
+                            targetAtSlot.grow(1);
                         }
+
+                        dispenserBlockEntity.setItem(randomSlot, item);
+                        return;
+                        // Kitin end
+
+                        // CraftBukkit start - Fire event when pushing items into other inventories
+//                        org.bukkit.craftbukkit.inventory.CraftItemStack oitemstack = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(item.copyWithCount(1));
+//
+//                        org.bukkit.inventory.Inventory destinationInventory;
+//                        // Have to special case large chests as they work oddly
+//                        if (containerAt instanceof net.minecraft.world.CompoundContainer compoundContainer) {
+//                            destinationInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest(compoundContainer);
+//                        } else {
+//                            destinationInventory = containerAt.getOwner().getInventory();
+//                        }
+//
+//                        org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(dispenserBlockEntity.getOwner().getInventory(), oitemstack, destinationInventory, true);
+//                        level.getCraftServer().getPluginManager().callEvent(event);
+//                        if (event.isCancelled()) {
+//                            return;
+//                        }
+//                        itemStack = HopperBlockEntity.addItem(dispenserBlockEntity, containerAt, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem()), direction.getOpposite());
+//                        if (event.getItem().equals(oitemstack) && itemStack.isEmpty()) {
+//                            // CraftBukkit end
+//
+//                            itemStack = item.copy();
+//                            itemStack.shrink(1);
+//                        } else {
+//                            itemStack = item.copy();
+//                        }
                     }
 
                     dispenserBlockEntity.setItem(randomSlot, itemStack);
