From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 23 Jan 2026 20:12:58 +0800
Subject: [PATCH] Network_globalMaxChunkSendRate


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index a36e9a35d8305406051d89bd3f9d703f07ffae59..ea7e980de0354caf59ca0a948c6cde1e5bf6dcb6 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -827,6 +827,11 @@ public final class RegionizedPlayerChunkLoader {
             final int maxSendsThisTick = Math.min((int)this.chunkSendLimiter.takeAllocation(time, sendRate, maxSends), this.sendQueue.size());
             // we do not return sends that we took from the allocation back because we want to limit the max send rate, not target it
             for (int i = 0; i < maxSendsThisTick; ++i) {
+                // Kitin - QoS globalMaxChunkSendRate
+                if (!me.sucixr.kitin.network.qos.GlobalChunkLimiter.tryAcquire()) {
+                    break;
+                }
+                // Kitin end
                 final long pendingSend = this.sendQueue.firstLong();
                 final int pendingSendX = CoordinateUtils.getChunkX(pendingSend);
                 final int pendingSendZ = CoordinateUtils.getChunkZ(pendingSend);
