From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Sun, 8 Feb 2026 01:29:33 +0800
Subject: [PATCH] VanillaFix_Login-State-NPE


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 10b159a61480b790a9fd320876ff731ab0897507..616b938a0f273cbd7c2bc47692860f07e42cfe53 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -260,16 +260,34 @@ public abstract class PlayerList {
             )
         );
         player.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
-        serverGamePacketListenerImpl.send(new ClientboundChangeDifficultyPacket(levelData.getDifficulty(), levelData.isDifficultyLocked()));
-        serverGamePacketListenerImpl.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
-        serverGamePacketListenerImpl.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot()));
-        RecipeManager recipeManager = this.server.getRecipeManager();
-        serverGamePacketListenerImpl.send(
-            new ClientboundUpdateRecipesPacket(recipeManager.getSynchronizedItemProperties(), recipeManager.getSynchronizedStonecutterRecipes())
-        );
-        this.sendPlayerPermissionLevel(player);
+//        serverGamePacketListenerImpl.send(new ClientboundChangeDifficultyPacket(levelData.getDifficulty(), levelData.isDifficultyLocked()));
+//        serverGamePacketListenerImpl.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
+//        serverGamePacketListenerImpl.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot())); // Kitin off The code has been migrated MojangFix_Login-State-NPE
+//        RecipeManager recipeManager = this.server.getRecipeManager();
+//        serverGamePacketListenerImpl.send(
+//            new ClientboundUpdateRecipesPacket(recipeManager.getSynchronizedItemProperties(), recipeManager.getSynchronizedStonecutterRecipes())
+//        );
+        //this.sendPlayerPermissionLevel(player); // Kitin off The code has been migrated MojangFix_Login-State-NPE
+
+        // Kitin start - MojangFix_Login-State-NPE
+        // 修复 Mod 客户端进服网络协议错误 (NPE) 问题
+        // 原因：Kitin核心优化了网络发包逻辑，导致状态包在客户端 World 初始化完成前到达。
+        // 原版也有这种问题，只是出现概率很低，因为会拼命发送区块包，自然导致玩家信息数据发送延迟
+        // 方案：延迟nTick发送
+        player.getBukkitEntity().taskScheduler.schedule((ServerPlayer p) -> {
+            if (p.connection == null) return;
+            p.connection.send(new ClientboundChangeDifficultyPacket(levelData.getDifficulty(), levelData.isDifficultyLocked()));
+            p.connection.send(new ClientboundPlayerAbilitiesPacket(p.getAbilities()));
+            p.connection.send(new ClientboundSetHeldSlotPacket(p.getInventory().getSelectedSlot()));
+            RecipeManager recipeManager = this.server.getRecipeManager();
+            p.connection.send(new ClientboundUpdateRecipesPacket(recipeManager.getSynchronizedItemProperties(), recipeManager.getSynchronizedStonecutterRecipes()));
+            this.sendPlayerPermissionLevel(p);
+            p.getRecipeBook().sendInitialRecipeBook(p);
+        }, null, 3L); // 延迟 3 tick
+        // Kitin end
+
         player.getStats().markAllDirty();
-        player.getRecipeBook().sendInitialRecipeBook(player);
+        //player.getRecipeBook().sendInitialRecipeBook(player); // Kitin off
         //this.updateEntireScoreboard(serverLevel.getScoreboard(), player); // Folia - region threading
         this.server.invalidateStatus();
         MutableComponent mutableComponent;
