From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:37:16 +0800
Subject: [PATCH] PaperFix_ItemEntity-Abnormal-movement-in-water


diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index b4f9f41a56bbc40e90217c0344c070ebdd08c119..cac3d4e2c589bf2fa9709efb7b21a0235b86e084 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -300,7 +300,7 @@ public final class ActivationRange {
      */
     public static boolean checkIfActive(final Entity entity) {
         // Never safe to skip fireworks or item gravity
-        if (entity instanceof FireworkRocketEntity || (entity instanceof ItemEntity && (entity.tickCount + entity.getId()) % 4 == 0)) { // Needed for item gravity, see ItemEntity tick
+        if (entity instanceof FireworkRocketEntity || ((entity instanceof ItemEntity && !entity.isInWater() && (entity.tickCount + entity.getId()) % 4 == 0))) { // Needed for item gravity, see ItemEntity tick // Kitin PaperFix_ItemEntity-Abnormal-movement-in-water
             return true;
         }
         // special case always immunities
