From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:36:22 +0800
Subject: [PATCH] 
 FoliaFix_ArrayIndexOutOfBoundsException-in-TransientCraftingContainer


diff --git a/net/minecraft/world/inventory/TransientCraftingContainer.java b/net/minecraft/world/inventory/TransientCraftingContainer.java
index 7263663bf06178d3eeed16155aef6acaa8726281..7d17650a234f035e3aef167a59a03c08618d5de8 100644
--- a/net/minecraft/world/inventory/TransientCraftingContainer.java
+++ b/net/minecraft/world/inventory/TransientCraftingContainer.java
@@ -14,7 +14,7 @@ public class TransientCraftingContainer implements CraftingContainer {
     private final AbstractContainerMenu menu;
 
     // CraftBukkit start - add fields
-    public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+    public List<org.bukkit.entity.HumanEntity> transaction = new java.util.concurrent.CopyOnWriteArrayList<>(); // Kitin FoliaFix_ArrayIndexOutOfBoundsException-in-TransientCraftingContainer
     private net.minecraft.world.item.crafting.@org.jspecify.annotations.Nullable RecipeHolder<net.minecraft.world.item.crafting.CraftingRecipe> currentRecipe;
     public net.minecraft.world.Container resultInventory;
     private Player owner;
