From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:41:07 +0800
Subject: [PATCH] Network_Staggered-Metadata-Sync


diff --git a/net/minecraft/network/syncher/SynchedEntityData.java b/net/minecraft/network/syncher/SynchedEntityData.java
index 48e3b67adf9c889363fed6a79c0eb7744806f681..bff468dae929d92c22aa44cef789ce2c14a2e017 100644
--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -260,4 +260,21 @@ public class SynchedEntityData {
             return new SynchedEntityData.DataValue<>(id, serializer, serializer.codec().decode(input));
         }
     }
+
+//    // Kitin start - Network_Staggered-Metadata-Sync
+//    public boolean kitin$hasCriticalDirtyData() {
+//        if (!this.isDirty) return false;
+//        boolean isPlayer = this.entity instanceof net.minecraft.world.entity.player.Player;
+//        for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
+//            if (dataItem.isDirty()) {
+//                int id = dataItem.getAccessor().id();
+//                if (isPlayer && (id == 10 || id == 17)) {
+//                    continue;
+//                }
+//                return true;
+//            }
+//        }
+//        return false;
+//    }
+//    // Kitin end
 }
diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 86e8cb60fedc7f046412e07708912521b61ac588..4b4d84eda77424a2f0abe9187273f4c36b617656 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -433,6 +433,17 @@ public class ServerEntity {
     private void sendDirtyEntityData() {
         SynchedEntityData entityData = this.entity.getEntityData();
         List<SynchedEntityData.DataValue<?>> packedValues = entityData.packDirty();
+
+//        // Kitin start - Network_Staggered-Metadata-Sync
+//        if (!entityData.isDirty()) {
+//            return;
+//        }
+//        boolean isStaggerTick = (this.tickCount + this.entity.getId()) % 20 == 0;
+//        if (!isStaggerTick && !entityData.kitin$hasCriticalDirtyData()) {
+//            return;
+//        }
+//        // Kitin end
+
         if (packedValues != null) {
             this.trackedDataValues = entityData.getNonDefaultValues();
             this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundSetEntityDataPacket(this.entity.getId(), packedValues));
