From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:54:07 +0800
Subject: [PATCH] Pref_Galaxy-Scheduler


diff --git a/io/papermc/paper/threadedregions/TickRegionScheduler.java b/io/papermc/paper/threadedregions/TickRegionScheduler.java
index 439c5a222fcf51c9dc5bfef9cdb23f419fffc28c..f7a4138c3cbc132e11bd1ae718db6cdcb40f3161 100644
--- a/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -53,6 +53,7 @@ public final class TickRegionScheduler {
 
     public static enum SchedulerType {
         EDF,
+        GALAXY, // Kitin Galaxy Scheduler
         WORK_STEALING;
     }
 
@@ -77,6 +78,12 @@ public final class TickRegionScheduler {
                 this.scheduler = new EDFSchedulerThreadPool(initialThreads, threadFactory);
                 break;
             }
+            // Kitin start - Kitin Galaxy Scheduler
+            case GALAXY: {
+                this.scheduler = new me.sucixr.kitin.performance.scheduler.galaxy.core.GalaxySchedulerThreadPool(initialThreads, threadFactory);
+                break;
+            }
+            // Kitin end
             case WORK_STEALING: {
                 this.scheduler = new StealingScheduledThreadPool(
                         threadFactory, MoonriseConstants.NUMA_ENABLE ? OSNuma.getNativeInstance() : OSNuma.NoOp.INSTANCE
@@ -94,6 +101,11 @@ public final class TickRegionScheduler {
         if (this.scheduler instanceof EDFSchedulerThreadPool edfSchedulerThreadPool) {
             edfSchedulerThreadPool.start();
         }
+        // Kitin start - Kitin Galaxy Scheduler
+        if (this.scheduler instanceof me.sucixr.kitin.performance.scheduler.galaxy.core.GalaxySchedulerThreadPool galaxyPool) {
+            galaxyPool.start();
+        }
+        // Kitin end
     }
 
     public void setThreads(final int threads) {
diff --git a/io/papermc/paper/threadedregions/TickRegions.java b/io/papermc/paper/threadedregions/TickRegions.java
index b75293fea5dc029374c1c03361e859f0ddb58e83..bf1a40bb9a969bc253f190b76005ca06386da8ae 100644
--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -45,11 +45,11 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
         int tickThreads;
         if (config.threads <= 0) {
             tickThreads = OSNuma.getNativeInstance().getTotalCores() / 2;
-            if (tickThreads <= 4) {
-                tickThreads = 1;
-            } else {
-                tickThreads =  tickThreads / 4;
-            }
+            // Kitin start
+            if (tickThreads <= 1) tickThreads = 1;
+            else if (tickThreads <=4) tickThreads = 2;
+            else tickThreads =  tickThreads / 2;
+            // Kitin end
         } else {
             tickThreads = config.threads;
         }
