From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Mon, 27 Jul 2026 18:14:03 +0800
Subject: [PATCH] Other_Luminol_Add-missing-teleportation-event-APIs-for-folia


diff --git a/src/main/java/io/papermc/paper/SparksFly.java b/src/main/java/io/papermc/paper/SparksFly.java
index 78d53b3887fa0e45cc8fa8d87041f03913b4f18b..8929b68f8ac4a3d9849252b17cdfab1c751b02b2 100644
--- a/src/main/java/io/papermc/paper/SparksFly.java
+++ b/src/main/java/io/papermc/paper/SparksFly.java
@@ -194,7 +194,7 @@ public final class SparksFly implements SparkPlugin { // Kitin
 
     @Override
     public String getVersion() {
-        return "Unknown-version-Kitin-Internal"; // Kitin - Hardcoded version
+        return "1.10.152-Kitin-Internal"; // Kitin - Hardcoded version
     }
 
     @Override
diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 532076c7358aa355ce5144075fe6bd79de6ef9e6..359fa0f7fb99f6b0d9999c864bcdfd59d5b78173 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -183,6 +183,12 @@ public final class MCUtil {
         return CraftLocation.toBukkit(pos, world);
     }
 
+    // Luminol start - add BlockPos support
+    public static Location toLocation(Level world, BlockPos pos) {
+        return new Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ());
+    }
+    // Luminol end - add BlockPos support
+
     /**
      * Converts a NMS World/Vector to Bukkit Location
      *
