From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:54:08 +0800
Subject: [PATCH] Pref_Galaxy-Scheduler


diff --git a/src/main/java/ca/spottedleaf/concurrentutil/scheduler/GalaxyTaskAccessor.java b/src/main/java/ca/spottedleaf/concurrentutil/scheduler/GalaxyTaskAccessor.java
new file mode 100644
index 0000000000000000000000000000000000000000..c94f9e04e2ecb63f9c292a1af6a5d7ae2aa72250
--- /dev/null
+++ b/src/main/java/ca/spottedleaf/concurrentutil/scheduler/GalaxyTaskAccessor.java
@@ -0,0 +1,17 @@
+package ca.spottedleaf.concurrentutil.scheduler;
+
+
+//Kitin Galaxy Scheduler 跳板类
+public final class GalaxyTaskAccessor {
+    public static Object getState(SchedulableTick tick) {
+        return tick.state;
+    }
+
+    public static void setState(SchedulableTick tick, Object state) {
+        tick.state = state;
+    }
+
+    public static long getScheduledStart(SchedulableTick tick) {
+        return tick.scheduledStart;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 10bec0f46f96605d82d8ec77327f2e8af18c0fd2..a0b0a0574665c1abc208e22831537239e8d8e1ce 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -401,10 +401,19 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         public int threads = -1;
         public int gridExponent = 4;
-        public io.papermc.paper.threadedregions.TickRegionScheduler.SchedulerType scheduler = io.papermc.paper.threadedregions.TickRegionScheduler.SchedulerType.EDF;
+        public io.papermc.paper.threadedregions.TickRegionScheduler.SchedulerType scheduler = io.papermc.paper.threadedregions.TickRegionScheduler.SchedulerType.GALAXY; // Kitin Galaxy Scheduler
+        public boolean alwaysUseKitinGalaxyScheduler = false; // Kitin Galaxy Scheduler // 暂时禁用
 
         @PostProcess
         public void postProcess() {
+            // Kitin start - Kitin Galaxy Scheduler
+            if (this.alwaysUseKitinGalaxyScheduler) {
+                if (this.scheduler != io.papermc.paper.threadedregions.TickRegionScheduler.SchedulerType.GALAXY) {
+                    LOGGER.warn("Default use Kitin Galaxy Scheduler, If you want to switch to another scheduler, please modify it in paper-global config");
+                    this.scheduler = io.papermc.paper.threadedregions.TickRegionScheduler.SchedulerType.GALAXY;
+                }
+            }
+            // Kitin end
             io.papermc.paper.threadedregions.TickRegions.init(this);
         }
     }
