From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 11:41:19 +0800
Subject: [PATCH] FoliaFix_ScheduleCommand


diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index d5ff949f4f459c4358b0130cd20f8ed9cf24b42d..6c554a6e3a79f5a87833640696dce09d57ed8e6f 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -32,6 +32,8 @@ import java.util.function.BooleanSupplier;
 
 public final class RegionizedServer {
 
+    public static volatile boolean IS_GLOBAL_TICKING = false; // Kitin FoliaFix_ScheduleCommand
+
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final RegionizedServer INSTANCE = new RegionizedServer();
 
@@ -331,6 +333,12 @@ public final class RegionizedServer {
         }
         this.randomWalk();
          */
+
+        // Kitin start - FoliaFix_ScheduleCommand
+        IS_GLOBAL_TICKING = true;
+        try {
+        // Kitin end
+
         ++this.tickCount;
         // expire invalid click command callbacks
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue((int)this.tickCount); // Paper // Folia - region threading - moved to global tick
@@ -345,6 +353,17 @@ public final class RegionizedServer {
         // commands
         ((DedicatedServer)MinecraftServer.getServer()).handleConsoleInputs();
 
+        //Kitin start - fix TimerQueue FoliaFix_ScheduleCommand use it
+        ServerLevel overworld = MinecraftServer.getServer().overworld();
+        if (overworld != null) {
+            overworld.serverLevelData.getGameTime(); // Paper - removed tick call
+        }
+        //Kitin end
+
+        // Kitin start - FoliaFix_ScheduleCommand
+        MinecraftServer.getServer().getFunctions().tick();
+        // Kitin end
+
         // needs
         // player ping sample
         // world global tick
@@ -363,6 +382,11 @@ public final class RegionizedServer {
 
         // player list
         MinecraftServer.getServer().getPlayerList().tick();
+        // Kitin start - FoliaFix_ScheduleCommand
+        } finally {
+            IS_GLOBAL_TICKING = false;
+        }
+        // Kitin end
     }
 
     private void tickPlayerSample() {
diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 1d7fb8ef08b0457be5c08cfb079cf2d7cb8bcc9f..131dc1b78a37ad0bb8942915a87a998b7a7daac3 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -239,7 +239,7 @@ public class Commands {
         RideCommand.register(this.dispatcher);
         RotateCommand.register(this.dispatcher);
         SayCommand.register(this.dispatcher);
-        //ScheduleCommand.register(this.dispatcher); // Folia - region threading
+        ScheduleCommand.register(this.dispatcher); // Folia - region threading //Kitin re
         //ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
         SeedCommand.register(this.dispatcher, commandSelection != Commands.CommandSelection.INTEGRATED);
         VersionCommand.register(this.dispatcher, commandSelection != Commands.CommandSelection.INTEGRATED);
diff --git a/net/minecraft/server/ServerFunctionManager.java b/net/minecraft/server/ServerFunctionManager.java
index a8aef0e78447520530e4c831ef198027b492e35d..ced7d335d31259edc2ef3b7a2c4d313b9693f226 100644
--- a/net/minecraft/server/ServerFunctionManager.java
+++ b/net/minecraft/server/ServerFunctionManager.java
@@ -66,9 +66,22 @@ public class ServerFunctionManager {
 
         try {
             InstantiatedFunction<CommandSourceStack> function = functionIn.instantiate(null, this.getDispatcher());
+            try{ // Kitin - FoliaFix_ScheduleCommand
             Commands.executeCommandInContext(
                 sender, context -> ExecutionContext.queueInitialFunctionCall(context, function, sender, CommandResultCallback.EMPTY)
             );
+            // Kitin start - FoliaFix_ScheduleCommand
+            }catch (RuntimeException e) {
+                // 如果捕获到在 Level.java 里抛出的特殊异常
+                if ("KITIN_SILENT_ABORT".equals(e.getMessage())) {
+                    // 直接吞掉！
+                    // 任务已经中止了（防闪烁），日志也不打印（防刷屏）。
+                    return;
+                }
+                // 其他异常照常抛出
+                throw e;
+            }
+            // Kitin end
         } catch (FunctionInstantiationException var9) {
         } catch (Exception var10) {
             LOGGER.warn("Failed to execute function {}", functionIn.id(), var10);
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 35305205047240dc91ed1bb60e742411a76e2a22..0b01d14f1a8d7957fb3fa7b900b888689dc4a991 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1352,8 +1352,9 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     @Override
     public BlockState getBlockState(final BlockPos pos) {
         // CraftBukkit start - tree generation
-        if (this.getCurrentWorldData().captureTreeGeneration) { // Folia - region threading
-            CraftBlockState previous = this.getCurrentWorldData().capturedBlockStates.get(pos); // Paper // Folia - region threading
+        var worldData = this.getCurrentWorldData(); // Kitin - FoliaFix_ScheduleCommand
+        if (worldData != null && worldData.captureTreeGeneration) { // Folia - region threading // Kitin - FoliaFix_ScheduleCommand
+            CraftBlockState previous = worldData.capturedBlockStates.get(pos); // Paper // Folia - region threading
             if (previous != null) {
                 return previous.getHandle();
             }
@@ -1850,6 +1851,23 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public <T extends Entity> void getEntities(final EntityTypeTest<Entity, T> entityTypeTest,
                                                final AABB boundingBox, final Predicate<? super T> predicate,
                                                final List<? super T> into, final int maxCount) {
+        // Kitin start - FoliaFix_ScheduleCommand
+        // 请注意，此修复并非完美修复
+        // 如果当前正在进行 Global Tick (由 schedule 触发)，并且试图查询实体，
+        // 我们直接抛出一个 RuntimeException。
+        // 1. 抛出异常 = 命令立即停止 = 此时"灭灯逻辑"还没跑 = 不会闪烁！
+        // 2. 这里的 check 在 ensureTickThread 之前 = 不会打印红字报错！
+        //方案阶段,处理方式,火把状态,后台状态,失败原因
+        //阶段 1,原版 Folia,不亮,无报错,全局线程被彻底禁用，没法干活。
+        //阶段 2,仅注入区域线程,亮,大量报错,绿轨干活了，但红轨撞墙了（被 Folia 拦截）。
+        //阶段 3,修改 Level 返回空,闪烁,无报错,红轨以为没玩家，主动把灯关了。
+        //阶段 4,最终方案 (静默熔断),常亮,无报错,绿轨干活，红轨在捣乱前被我们“暗杀”了。
+        if (io.papermc.paper.threadedregions.RegionizedServer.IS_GLOBAL_TICKING) {
+            //throw new RuntimeException("KITIN_SILENT_ABORT");
+            return; //无意义报错，这个是修动态光源数据包用的
+        }
+        // Kitin end
+
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, boundingBox, "Cannot getEntities asynchronously"); // Folia - region threading
         Profiler.get().incrementCounter("getEntities");
 
