From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SucIXR <sasasa123456@foxmail.com>
Date: Fri, 24 Jul 2026 12:08:10 +0800
Subject: [PATCH] VanillaFix_Login-State-NPE


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 4da7d5201ede250de14e6e6004bcc1230b67427f..ad5b1b827a74b0171da8a2ce7e41268d10547d20 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -262,16 +262,34 @@ public abstract class PlayerList {
             )
         );
         player.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
-        playerConnection.send(new ClientboundChangeDifficultyPacket(levelData.getDifficulty(), levelData.isDifficultyLocked()));
-        playerConnection.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
-        playerConnection.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot()));
-        RecipeManager recipeManager = this.server.getRecipeManager();
-        playerConnection.send(
-            new ClientboundUpdateRecipesPacket(recipeManager.getSynchronizedItemProperties(), recipeManager.getSynchronizedStonecutterRecipes())
-        );
-        this.sendPlayerPermissionLevel(player);
+//        playerConnection.send(new ClientboundChangeDifficultyPacket(levelData.getDifficulty(), levelData.isDifficultyLocked()));
+//        playerConnection.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
+//        playerConnection.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot()));
+//        RecipeManager recipeManager = this.server.getRecipeManager();
+//        playerConnection.send(
+//            new ClientboundUpdateRecipesPacket(recipeManager.getSynchronizedItemProperties(), recipeManager.getSynchronizedStonecutterRecipes())
+//        );
+//        this.sendPlayerPermissionLevel(player); // Kitin off
+
+        // Kitin start - MojangFix_Login-State-NPE
+        // 修复 Mod 客户端进服网络协议错误 (NPE) 问题
+        // 原因：Kitin核心优化了网络发包逻辑，导致状态包在客户端 World 初始化完成前到达。
+        // 原版也有这种问题，只是出现概率很低，因为会拼命发送区块包，自然导致玩家信息数据发送延迟
+        // 方案：延迟nTick发送
+        player.getBukkitEntity().taskScheduler.schedule((ServerPlayer p) -> {
+            if (p.connection == null) return;
+            p.connection.send(new ClientboundChangeDifficultyPacket(levelData.getDifficulty(), levelData.isDifficultyLocked()));
+            p.connection.send(new ClientboundPlayerAbilitiesPacket(p.getAbilities()));
+            p.connection.send(new ClientboundSetHeldSlotPacket(p.getInventory().getSelectedSlot()));
+            RecipeManager recipeManager = this.server.getRecipeManager();
+            p.connection.send(new ClientboundUpdateRecipesPacket(recipeManager.getSynchronizedItemProperties(), recipeManager.getSynchronizedStonecutterRecipes()));
+            this.sendPlayerPermissionLevel(p);
+            p.getRecipeBook().sendInitialRecipeBook(p);
+        }, null, 3L); // 延迟 3 tick
+        // Kitin end
+
         player.getStats().markAllDirty();
-        player.getRecipeBook().sendInitialRecipeBook(player);
+        //player.getRecipeBook().sendInitialRecipeBook(player); // Kitin off
         //this.updateEntireScoreboard(level.getScoreboard(), player); // Folia - region threading
         this.server.invalidateStatus();
         MutableComponent component;
